### PR TITLE
EOS-27028: Spiel APIs for retriving byte count and status of pool version

### PR DESCRIPTION
**Pull request by AbhishekSahaSeagate**
Changes:
- Added spiel API m0_spiel_conf_pver_status() which returns the status (HEALTHY|DEGRADED|CRITICAL|DAMAGED).
- Added spiel stub API m0_spiel_proc_counters_fetch() to fetch the bytecount data from a given ioservice.
- Added a spiel UT and modified motr_utils_st.sh to test these APIs.

Signed-off-by: Abhishek Saha <abhishek.saha@seagate.com>

# Problem Statement
- Create Stub spiel APIs for retriving bytecount from ioservice and getting current failure status of a given pool version.

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: motr/st/utils/motr_utils_st.sh [hunk 1]
```diff
@@ -112,193 +112,193 @@ test_with_N_K()
...
-	send -- "read $object_id2 $dest_file $block_size $block_count $blks_per_io\r"
-	expect "m0client >>"
-	send -- "delete $object_id3\r"
-	expect "m0client >>"
-	send -- "delete $object_id2\r"
-	expect "m0client >>"
-	send -- "quit\r"
-EOF
-	echo "m0client test is Successful"
+#	# Test m0client utility
```
Comment (madhavemuri) [inline motr/st/utils/motr_utils_st.sh:115]: Why this code is commented ?

Comment (AbhishekSahaSeagate) [reply motr/st/utils/motr_utils_st.sh:115]: I have used this ST to test out the m0_conf_pver_status() API, I just wanted to run m0cp once to see the output, so commented rest of the ST. This change is not intended to be merged.

**Comment by mssawant** [comment]
@AbhishekSahaSeagate, I think we also need a corresponding spiel interface to fetch a pool versions status that takes spiel instance as an input, check the spiel interface to fetch filesystem stats,
```
M0_INTERNAL int m0_spiel__fs_stats_fetch(struct m0_spiel_core *spc,
struct m0_fs_stats *stats)
{
struct _fs_stats_ctx fsx;
struct spiel_proc_item *proc;
int rc;

M0_ENTRY();
M0_PRE(spc != NULL);
M0_PRE(spc->spc_confc != NULL);
M0_PRE(stats != NULL);

if (!m0_confc_is_inited(spc->spc_confc)) <<<<<<<<<<<<<<
return M0_ERR_INFO(-EAGAIN, "confc is finalised");

spiel__fs_stats_ctx_init(&fsx, spc, &M0_CONF_PROCESS_TYPE);
/* walk along the filesystem nodes and get to process level */
rc = SPIEL_CONF_DIR_ITERATE(spc->spc_confc, &fsx,
spiel__item_enlist,
M0_CONF_ROOT_NODES_FID,
M0_CONF_NODE_PROCESSES_FID) ?:
```
cc @Shreya-18, @madhavemuri

**Comment by madhavemuri** [comment]
> @AbhishekSahaSeagate, I think we also need a corresponding spiel interface to fetch a pool version's status that takes spiel instance as an input, check the spiel interface to fetch filesystem stats,
> 
> ```
> M0_INTERNAL int m0_spiel__fs_stats_fetch(struct m0_spiel_core *spc,
> struct m0_fs_stats *stats)
> {
> struct _fs_stats_ctx fsx;
> struct spiel_proc_item *proc;
> int rc;
> 
> M0_ENTRY();
> M0_PRE(spc != NULL);
> M0_PRE(spc->spc_confc != NULL);
> M0_PRE(stats != NULL);
> 
> if (!m0_confc_is_inited(spc->spc_confc)) <<<<<<<<<<<<<<
> return M0_ERR_INFO(-EAGAIN, "confc is finalised");
> 
> spiel__fs_stats_ctx_init(&fsx, spc, &M0_CONF_PROCESS_TYPE);
> /* walk along the filesystem nodes and get to process level */
> rc = SPIEL_CONF_DIR_ITERATE(spc->spc_confc, &fsx,
> spiel__item_enlist,
> M0_CONF_ROOT_NODES_FID,
> M0_CONF_NODE_PROCESSES_FID) ?:
> ```
> 
> cc @Shreya-18, @madhavemuri

@AbhishekSahaSeagate : Let's add spiel API for it assuming spiel confc is already inited.

File: conf/pvers.c [hunk 2]
```diff
@@ -804,6 +804,52 @@ static int conf_pver_virtual_create(const struct m0_fid *fid,
...
+		out_info->cpi_state = M0_CPS_HEALTHY;
+	else if (failures < K)
+		out_info->cpi_state = M0_CPS_DEGRADED;
+	else if (failures == K)
+		out_info->cpi_state = M0_CPS_CRITICAL;
+	else
+		out_info->cpi_state = M0_CPS_DAMAGED;
+
+	return M0_RC(rc);
+}M0_EXPORTED(m0_conf_pver_status);
```
Comment (madhavemuri) [inline conf/pvers.c:851]: Add space between } and M0_EXPORTED()

File: conf/pvers.h [hunk 3]
```diff
@@ -106,6 +107,44 @@
  * @{
  */
 
+enum m0_conf_pver_state {
+	M0_CPS_HEALTHY,
```
Comment (madhavemuri) [inline conf/pvers.h:111]: Add comments about these states here.

File: spiel/cmd.c [hunk 4]
```diff
@@ -1655,6 +1655,31 @@ int m0_spiel_pool_rebalance_abort(struct m0_spiel     *spl,
...
+ * List element of an individual key value pair on the byte count btree.
+ */
+struct m0_proc_data {
+	struct m0_spiel_bckey pd_bckey;
+	struct m0_spiel_bcrec pd_bcrec;
+	struct m0_tlink       pd_link;
+};
+
+/**
+ * A context for fetching counters from ioservices.
```
Comment (madhavemuri) [inline spiel/cmd.c:1672]: Rename ioservice to motr servers.

File: spiel/cmd.c [hunk 5]
```diff
@@ -2082,6 +2135,25 @@ int m0_spiel_confstr(struct m0_spiel *spl, char **out)
...
+	int              rc;
+
+	M0_ENTRY();
+	M0_PRE(spl != NULL);
+
+	confc = spl->spl_core.spc_confc;
+	if (!m0_confc_is_inited(confc))
+		return M0_ERR_INFO(-EAGAIN, "confc is finalised");
+
+	rc = m0_conf_pver_status(fid, confc, out_info);
```
Comment (madhavemuri) [inline spiel/cmd.c:2152]: return M0_RC(m0_conf_pver_status(fid, confc, out_info));

File: spiel/spiel.h [hunk 6]
```diff
@@ -1249,6 +1250,54 @@ int m0_spiel_dix_rebalance_abort(struct m0_spiel     *spl,
 int m0_spiel_pool_rebalance_abort(struct m0_spiel     *spl,
 			          const struct m0_fid *pool_fid);
 
+/**
+ * Motr Byte count stats. These stats contain byte count and object count data
```
Comment (madhavemuri) [inline spiel/spiel.h:1254]: s/object count/motr object count

File: spiel/spiel.h [hunk 7]
```diff
@@ -1249,6 +1250,54 @@ int m0_spiel_dix_rebalance_abort(struct m0_spiel     *spl,
...
+	int                     pc_rc;
+	uint32_t                pc_cnt; /** Length of Key and Value arrays */
+	struct m0_fid           pc_proc_fid;
+	struct m0_spiel_bckey **pc_bckey;
+	struct m0_spiel_bcrec **pc_bcrec;
+};
+
+/**
+ * Fetches byte counts and object counts per pool version written by all
+ * end users from a target ioservice. All entires in the byte count btree
```
Comment (madhavemuri) [inline spiel/spiel.h:1287]: s/entires/entries

**Review by madhavemuri — APPROVED** [review]

File: conf/pvers.c (hunk 2 above)
Comment (AbhishekSahaSeagate) [reply conf/pvers.c:851]: Done

File: conf/pvers.h (hunk 3 above)
Comment (AbhishekSahaSeagate) [reply conf/pvers.h:111]: Done.

File: spiel/cmd.c (hunk 4 above)
Comment (AbhishekSahaSeagate) [reply spiel/cmd.c:1672]: Done.

File: spiel/cmd.c (hunk 5 above)
Comment (AbhishekSahaSeagate) [reply spiel/cmd.c:2152]: Done.

File: spiel/spiel.h (hunk 6 above)
Comment (AbhishekSahaSeagate) [reply spiel/spiel.h:1254]: Done.

File: spiel/spiel.h (hunk 7 above)
Comment (AbhishekSahaSeagate) [reply spiel/spiel.h:1287]: Done.

**Pull request merged**